### PR TITLE
INT-6770 - Query VMs by host

### PR DESCRIPTION
**Pull request by adam-in-ict**
# Description

Thank you for contributing to a JupiterOne integration!

Please include a summary of the change and which issue is fixed. Please also
include relevant motivation and context. List any dependencies that are required
for this change.

## Summary

Updating the query that lists VMs to query by host.  This reduces the likelihood that we will hit the API limit of 4000 VMs returned by that endpoint.

## Type of change

Please leave any irrelevant options unchecked.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [ ] This change requires a documentation update

### General Development Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### Integration Development Checklist:

Please leave any irrelevant options unchecked.

- [ ] I have checked for additional permissions required to call any new API
      endpoints, and have documented any additional permissions in
      `jupiterone.md`, where necessary.
- [ ] My changes properly paginate the target service provider's API
- [ ] My changes properly handle rate limiting of the target service provider's
      API
- [ ] My new integration step is instrumented to execute in the correct order
      using `dependsOn`
- [ ] I have referred to the
      [JupiterOne data model](https://github.com/JupiterOne/data-model/tree/main/src/schemas)
      to ensure that any new entities/relationships, and relevant properties,
      match the recommended model for this class of data
- [ ] I have updated the `CHANGELOG.md` file to describe my changes
- [ ] When changes include modifications to existing graph data ingestion, I've
      reviewed all existing managed questions referencing the entities,
      relationships, and their property names, to ensure those questions still
      function with my changes.

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -55,12 +55,22 @@ export class APIClient {
   private withBaseUri = (path: string) => `${this.baseUri}${path}`;
   private sessionId = '';
 
+  private async version702orNewer() {
```
Comment (adam-in-ict) [inline src/client.ts:58]: Adding a few helper functions for our version checks.  We may still want to revamp the entire client, but this at least makes it more readable for now.

File: src/steps/network/index.ts [hunk 2]
```diff
@@ -34,6 +34,11 @@ export async function buildVmNetworkRelationship({
     async (vmEntity) => {
       const vm = await apiClient.getVm(vmEntity.vm as string);
       for (const nics of Object.values(vm.nics)) {
+        // Depending on version, we may need to slightly modify where
```
Comment (adam-in-ict) [inline src/steps/network/index.ts:37]: Bonus change.  Unrelated to the VM issues we've been seeing, but during development I realized the output format is different enough that this step will fail without this change.

File: src/steps/virtual-machine/index.ts [hunk 3]
```diff
@@ -20,21 +20,27 @@ export async function fetchVms({
...
-    } catch (err) {
-      logger.info(
-        `Unable to query vcenter/vm/${vm.vm}/guest/identity endpoint.  This may be due to permissions or having VMware tools turned off on that VM.`,
-      );
-      vmGuest = null;
-      guestQueryFailCount++;
-    }
-    await jobState.addEntity(createVmEntity(vm, vmGuest));
-  });
+  await jobState.iterateEntities(
```
Comment (adam-in-ict) [inline src/steps/virtual-machine/index.ts:23]: Main change.  We're wrapping our old call to iterateVms in an iterateEntities call that will get all of our hosts.  This step was already dependent on hosts, so this doesn't really have any negative impacts that I've been able to find.

**Review by zemberdotnet — APPROVED** [review]

File: src/steps/virtual-machine/index.ts [hunk 4]
```diff
@@ -20,21 +20,27 @@ export async function fetchVms({
...
+          guestQuerySuccessCount++;
+        } catch (err) {
+          logger.info(
+            `Unable to query vcenter/vm/${vm.vm}/guest/identity endpoint.  This may be due to permissions or having VMware tools turned off on that VM.`,
+          );
+          vmGuest = null;
+          guestQueryFailCount++;
+        }
+        await jobState.addEntity(createVmEntity(vm, vmGuest));
+      }, hostEntity.hostname as string);
```
Comment (VDubber) [inline src/steps/virtual-machine/index.ts:41]: nit: hostname gets lost being after the lambda.

Comment (adam-in-ict) [reply src/steps/virtual-machine/index.ts:41]: Good call out.  I tend to agree that it's a nicer pattern to put the lambda last.

I'll get it updated in the next PR.  Thanks!

**Pull request merged**